uefi-test-runner/src/proto/network/snp.rs

```diff
@@ -0,0 +1,128 @@
+
+    let handles = bt.find_handles::<SimpleNetwork>().unwrap_or_default();
+
+    for handle in handles {
+        let simple_network = bt.open_protocol_exclusive::<SimpleNetwork>(handle);
+        if simple_network.is_err() {
+            continue;
+        }
+        let simple_network = simple_network.unwrap();
+
+        // Check shutdown
+        simple_network
+            .shutdown()
+            .expect("Failed to shutdown Simple Network");
+
+        // Check stop
+        simple_network
+            .stop()
+            .expect("Failed to stop Simple Network");
+
+        // Check start
+        simple_network
+            .start()
+            .expect("Failed to start Simple Network");
+
+        // Check initialize
+        simple_network
+            .initialize(0, 0)
+            .expect("Failed to initialize Simple Network");
+
+        simple_network.reset_statistics().unwrap();
+
+        // Reading the interrupt status clears it
+        simple_network.get_interrupt_status().unwrap();
+
+        // Set receive filters
+        simple_network
+            .receive_filters(
+                EFI_SIMPLE_NETWORK_RECEIVE_UNICAST
+                    | EFI_SIMPLE_NETWORK_RECEIVE_MULTICAST
+                    | EFI_SIMPLE_NETWORK_RECEIVE_BROADCAST,
+                0,
+                false,
+                None,
+            )
+            .expect("Failed to set receive filters");
+
+        // Check media
+        if !simple_network.mode().media_present_supported || !simple_network.mode().media_present {
+            continue;
+        }
+
+        let payload = b"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\
+            \x45\x00\
+            \x00\x21\
+            \x00\x01\
+            \x00\x00\
+            \x10\
+            \x11\
+            \x07\x6a\
+            \xc0\xa8\x11\x0f\
+            \xc0\xa8\x11\x02\
+            \x54\x45\
+            \x54\x44\
+            \x00\x0d\
+            \xa9\xe4\
+            \x04\x01\x02\x03\x04";
+
+        let dest_addr = MacAddress([0xffu8; 32]);
+        assert!(!simple_network
+            .get_interrupt_status()
+            .unwrap()
+            .transmit_interrupt());
+        // Send the frame
+        simple_network
+            .transmit(
+                simple_network.mode().media_header_size as usize,
+                payload,
+                None,
+                Some(&dest_addr),
+                Some(&0x0800),
+            )
+            .expect("Failed to transmit frame");
+
+        info!("Waiting for the transmit");
+        while !simple_network
+            .get_interrupt_status()
+            .unwrap()
+            .transmit_interrupt()
+        {}
+
+        // Attempt to receive a frame
+        let mut buffer = [0u8; 1500];
+
+        info!("Waiting for the reception");
+        if simple_network.receive(&mut buffer, None, None, None, None)
+            == Err(Status::NOT_READY.into())
+        {
+            bt.stall(1_000_000);
```

Comment:
Ideally we could use the `wait_for_packet` here. I recall from the last PR attempting this that we couldn't get that event to fire... any idea what the problem could be there?

Reply:
I dug around in the EDK2 and OVMF source a bit but don't have much to add. It does look like the `WaitForPacket` Event and all related machinery have been implemented for at least the `virtio-net-pci` driver in QEMU. However, even while using this driver, the event never seems to be signaled. 

I'll try to plug gdb into qemu tomorrow and see what might be going on.

Reply:
> This is also why the event is not exposed, since I haven't managed to get it to work :)

I would recommend exposing it anyway; it may well work on some non-QEMU/EDK2 environments.

Reply:
I exposed it with a comment that cautions the user to check before using it :)